pkg/linkcache/cache.go

```diff
@@ -0,0 +1,193 @@
+package linkcache
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+var partitionNameRegex = regexp.MustCompile(`-part[0-9]+$`)
+
+// fsInterface defines the filesystem operations needed by ListingCache
+type fsInterface interface {
+	ReadDir(name string) ([]os.DirEntry, error)
+	EvalSymlinks(path string) (string, error)
+}
+
+// realFS implements fsInterface using the real filesystem
+type realFS struct{}
+
+func (f *realFS) ReadDir(name string) ([]os.DirEntry, error) {
+	return os.ReadDir(name)
+}
+
+func (f *realFS) EvalSymlinks(path string) (string, error) {
+	return filepath.EvalSymlinks(path)
+}
+
+// ListingCache polls the filesystem at the specified directory once per
+// period and checks each non-directory entry for a symlink. The results are
+// cached. Changes to the cache are logged, as well as the full contents of the
+// cache. The cache's Run() method is expected to be called in a goroutine. Its
+// cancellation is controlled via the context argument.
+type ListingCache struct {
+	period time.Duration
+	dir    string
+	links  *linkCache
+	fs     fsInterface
+}
+
+func NewListingCache(period time.Duration, dir string) *ListingCache {
+	return &ListingCache{
+		period: period,
+		dir:    dir,
+		links:  newLinkCache(),
+		fs:     &realFS{},
+	}
+}
+
+// Run starts the cache's background loop. The filesystem is listed and the cache
+// updated according to the frequency specified by the period. It will run until
+// the context is cancelled.
+func (l *ListingCache) Run(ctx context.Context) {
+	// Start the loop that runs every minute
+	ticker := time.NewTicker(l.period)
+	defer ticker.Stop()
+
+	// Initial list and update so we don't wait for the first tick.
+	err := l.listAndUpdate()
+	if err != nil {
+		klog.Warningf("Error listing and updating symlinks: %v", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			klog.Infof("Context done, stopping watcher")
+			return
+		case <-ticker.C:
+			err := l.listAndUpdate()
+			if err != nil {
+				klog.Warningf("Error listing and updating symlinks: %v", err)
+				continue
+			}
+
+			klog.Infof("periodic symlink cache read: %s", l.links.String())
+		}
+	}
+}
+
+func (l *ListingCache) listAndUpdate() error {
```

Comment:
How do normal attach/detach flows show up with this? When we log, do we have a way to distinguish between normal attach/detach sequence vs when the device mapping unexpectantly changes post-mount?

I'm concerned about normal flows creating too much noise in the logs and drowning out the real issues.